### PR TITLE
JDK-8219917: [WebView] Sub-resource integrity check fails

### LINKED ISSUE #230: JDK-8219917: [WebView] Sub-resource integrity check fails on Windows and Linux
I am unable to load a site using `WebView`. The specific site I'm trying on is Microsoft's oauth site. The error from the web console is:
```
Console: [null:0] Cannot load stylesheet https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/converged.v2.login.min_t7iocdq0wq2qh0nv233jig2.css. Failed integrity metadata check.
Console: [https://login.microsoftonline.com/common/oauth2/v2.0/authorize?.... Cannot load stylesheet https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/converged.v2.login.min_t7iocdq0wq2qh0nv233jig2.css. Failed integrity metadata check.
Console: [https://login.microsoftonline.com/common/oauth2/v2.0/authorize?.... Cannot load stylesheet https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/converged.v2.login.min_t7iocdq0wq2qh0nv233jig2.css. Failed integrity metadata check.
Console: [https://login.microsoftonline.com/common/oauth2/v2.0/authorize?.... Failed to load external resource ['https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/converged.v2.login.min_t7iocdq0wq2qh0nv233jig2.css']
Console: [null:0] Cannot load script https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/oldconvergedlogin_pcore.min_ueaoy8qjjryroqi9aubaoa2.js. Failed integrity metadata check.
Console: [null:0] Cannot load script https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/convergedloginpaginatedstrings-en.min_9eyqxjb02ixrn2nkbabbcq2.js. Failed integrity metadata check.
Console: [null:0] Cannot load script https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/oldconvergedlogin_pcore.min_ueaoy8qjjryroqi9aubaoa2.js. Failed integrity metadata check.
Console: [null:0] Cannot load script https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/convergedloginpaginatedstrings-en.min_9eyqxjb02ixrn2nkbabbcq2.js. Failed integrity metadata check.
Console: [null:0] Cannot load script https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/oldconvergedlogin_pcore.min_ueaoy8qjjryroqi9aubaoa2.js. Failed integrity metadata check.
Console: [https://login.microsoftonline.com/common/oauth2/v2.0/authorize?.... Failed to load external resource ['https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/oldconvergedlogin_pcore.min_ueaoy8qjjryroqi9aubaoa2.js']
Console: [null:0] Cannot load script https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/convergedloginpaginatedstrings-en.min_9eyqxjb02ixrn2nkbabbcq2.js. Failed integrity metadata check.
Console: [https://login.microsoftonline.com/common/oauth2/v2.0/authorize?.... Failed to load external resource ['https://secure.aadcdn.microsoftonline-p.com/ests/2.1.8148.16/content/cdnbundles/convergedloginpaginatedstrings-en.min_9eyqxjb02ixrn2nkbabbcq2.js']
```
(params replaced with `...`)
This happened with JavaFX 8 and after reworking with OpenJFX 11, the same error occurred.

I tried adding `sun.net.http.allowRestrictedHeaders=true` as several online similar issues suggested, but that did not affect the errors.

The System Properties (in-runtime printout):
```
{sun.desktop=windows, awt.toolkit=sun.awt.windows.WToolkit, java.specification.version=10, file.encoding.pkg=sun.io, sun.cpu.isalist=amd64, sun.jnu.encoding=Cp1252, java.class.path=C:\dev\src\github.com\erichiller\marshal-kotlin\build\classes\kotlin\main;C:\dev\lib\java\tornadofx-1.7.18-SNAPSHOT.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-jdk8\1.2.70\17bdfaa19afad44a641a778a1cacc53b0e95dddc\kotlin-stdlib-jdk8-1.2.70.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.apache.lucene\lucene-analyzers-common\7.4.0\e1afb580df500626a1c695e0fc9a7e8a8f58bcac\lucene-analyzers-common-7.4.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.apache.lucene\lucene-codecs\7.4.0\a9da1b39322df815da029d656baad7fa4e037dce\lucene-codecs-7.4.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.apache.lucene\lucene-queryparser\7.4.0\8e58add0d0c39df97d07c8e343041989bf4b3a3f\lucene-queryparser-7.4.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.apache.lucene\lucene-core\7.4.0\730d9ac80436c8cbc0b2a8a749259be536b97316\lucene-core-7.4.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\com.github.kittinunf.fuel\fuel\1.15.0\aefaf63dbea0b9d83b161728a4540f6463e3bd22\fuel-1.15.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-web\11\5c8a3b6cacde832437dff8e21a82749b5f66c1cf\javafx-web-11-win.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-controls\11\af8d24de44e16a19ae959028eb23a48c145fb682\javafx-controls-11-win.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-controls\11\58d961774262ec972bf304e16c154a8e18c2050b\javafx-controls-11.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-media\11\b7a33f3ff0cad17af71e883c0921bc6819cad9c8\javafx-media-11-win.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-media\11\586c51942ea38145727e345387b6b7cd53de0b33\javafx-media-11.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-graphics\11\9345c7ea43dbcc73a62eca5332717b4ce617e6c6\javafx-graphics-11-win.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-graphics\11\a736dd079047ec0b72b8c4970842a5c5e0c19f2f\javafx-graphics-11.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-base\11\ad3a4acfaa28b484d30c40c1f009445cc3f6e68b\javafx-base-11-win.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.openjfx\javafx-base\11\9fcd3e8e3227ec97bf503f7991fad1f3b14d005\javafx-base-11.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-jdk7\1.2.70\9318bb9dadaa56534ac8b0728a1ffa2bd0e48b28\kotlin-stdlib-jdk7-1.2.70.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\com.github.kittinunf.result\result\1.5.0\1be1ca0a17fe24ac18716e8193c7f38e847c4212\result-1.5.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib\1.2.70\b5b9449f73ce7bf312e89a7560cb3c209a0fa13e\kotlin-stdlib-1.2.70.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.apache.lucene\lucene-queries\7.4.0\229a50e6d9d4db076f671c230d493000c6e2972c\lucene-queries-7.4.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.apache.lucene\lucene-sandbox\7.4.0\1692604fa06a945d1ee19939022ef1a912235db3\lucene-sandbox-7.4.0.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-common\1.2.70\6ea722d7d0b9451293e739076ddb6a4108d57b6b\kotlin-stdlib-common-1.2.70.jar;C:\dev\conf\.gradle\caches\modules-2\files-2.1\org.jetbrains\annotations\13.0\919f0dfe192fb4e063e7dacadee7f8bb9a2672a9\annotations-13.0.jar, java.vm.vendor="Oracle Corporation", sun.arch.data.model=64, user.variant=, java.vendor.url=http://java.oracle.com/, user.timezone=, os.name=Windows 10, java.vm.specification.version=10, sun.java.launcher=SUN_STANDARD, user.country=US, sun.boot.library.path=C:\dev\lib\jdk\jdk-10.0.2\bin, sun.java.command=ai.marshal.desktop.AppKt, jdk.debug=release, sun.cpu.endian=little, user.home=C:\Users\eric, user.language=en, java.specification.vendor=Oracle Corporation, java.version.date=2018-07-17, java.home=C:\dev\lib\jdk\jdk-10.0.2, file.separator=\, java.vm.compressedOopsMode=Zero based, line.separator=
, java.specification.name=Java Platform API Specification, java.vm.specification.vendor=Oracle Corporation, java.awt.graphicsenv=sun.awt.Win32GraphicsEnvironment, user.script=, sun.management.compiler=HotSpot 64-Bit Tiered Compilers, java.runtime.version=10.0.2+13, user.name=eric, path.separator=;, os.version=10.0, java.runtime.name=OpenJDK Runtime Environment, file.encoding=UTF-8, sun.net.http.allowRestrictedHeaders=true, java.vm.name=OpenJDK 64-Bit Server VM, java.vendor.version=18.3, java.vendor.url.bug=http://bugreport.java.com/bugreport/, java.io.tmpdir=C:\Users\eric\AppData\Local\Temp\, java.version=10.0.2, user.dir=C:\dev\src\github.com\erichiller\marshal-kotlin, os.arch=amd64, java.vm.specification.name=Java Virtual Machine Specification, java.awt.printerjob=sun.awt.windows.WPrinterJob, sun.os.patch.level=, java.library.path=C:\dev\lib\jdk\jdk-10.0.2\bin;C:\Windows\Sun\Java\bin;C:\Windows\system32;C:\Windows;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\WindowsPowerShell\Modules\omega\system\git\cmd;C:\Program Files\WindowsPowerShell\Modules\omega\system\python\;C:\Program Files\WindowsPowerShell\Modules\omega\system\python\Scripts;C:\Program Files\WindowsPowerShell\Modules\omega\bin;C:\Program Files\WindowsPowerShell\Modules\omega\system\nodejs;C:\Program Files\WindowsPowerShell\Modules\omega\system\vim;C:\dev\lib\javafx-sdk-11\bin;C:\Users\eric\AppData\Local\Microsoft\WindowsApps;;C:\Users\eric\AppData\Local\Programs\Microsoft VS Code\bin;., java.vendor=Oracle Corporation, java.vm.info=mixed mode, java.vm.version=10.0.2+13, sun.io.unicode.encoding=UnicodeLittle, java.class.version=54.0}
```

JavaVM is 10.0.2

I am relatively sure that the issue is with CORS (I am loading `microsoftonline.com` and the resources are on `microsoftonline-p.com`). I've tried every possible fix that I can think of, or find online.

I have tried setting all of these

    engine.setJavaScriptEnabled(true)

    engine.setUserAgent("AppleWebKit/537.44")

    System.setProperty("sun.net.http.allowRestrictedHeaders", "true")

I've also set the Property via `-Dsun.net.http.allowRestrictedHeaders=true` in `VMOptions` ( as seen [here](https://stackoverflow.com/questions/44905264/cannot-sign-in-to-google-in-javafx-webview) ) and from the JavaFX issue [JDK-8096797](https://bugs.openjdk.java.net/browse/JDK-8096797)

The property shows up as set:

    println(System.getProperty("sun.net.http.allowRestrictedHeaders"))

yields `true` on the console printout.

Still no change, ever in the page output, it is always a white, blank screen, and the same errors from the web console.

I think I found the [trouble lines in WebKit(see `matchIntegrityMetadata`)](https://github.com/WebKit/webkit/blob/89c28d471fae35f1788a0f857067896a10af8974/Source/WebCore/loader/SubresourceIntegrity.cpp) even, but that didn't help me solve the problem any, because I don't know how to disable the _integrityCheck_.

For reference, here is the entire method:

    private fun WebView.authWindow(provider: Oath2Account){
    	engine.setUserAgent("AppleWebKit/537.44")
    	engine.setJavaScriptEnabled(true)
    	URLPermission("https://*.com")
    	System.setProperty("sun.net.http.allowRestrictedHeaders", "true")
    
    	Platform.runLater {
    		engine.userDataDirectory = File("C:\\users\\eric\\javafx_tmp")
    		engine.setOnError { println("IN PAGE ERROR --> $it") }
    		engine.setOnAlert { println("IN PAGE ALERT --> $it") }
    		engine.setConfirmHandler { println("IN PAGE CONFIRM HANDLER --> $it")
    		true
    		}
    		engine.setCreatePopupHandler { println("IN PAGE POPUP --> $it")
    		engine}
    		engine.setOnResized {   println("IN PAGE RESIZED --> $it") }
    		engine.setOnStatusChanged {   println("IN PAGE STATUS CHANGED --> $it")
    			println("\t${it.data}")
    			println("\t${it.source}")
    			println("\t${it.eventType}")
    			println("\t${it.target}")
    			println("\t${it.isConsumed}")
    
    		}
    		engine.setOnVisibilityChanged { println("IN PAGE VISIBILITY CHANGED --> $it") }
    		engine.setPromptHandler { println("IN PAGE PROMPTED --> $it")
    		"HELLO"}
    		println("JavaScript engine status: ${engine.isJavaScriptEnabled}")
    
    		println("engine is loading $loadURL")
    		engine.locationProperty().addListener { observable, oldLocation, newLocation->
    			println("observable=$observable\noldLocation=$oldLocation\nnewLocation=$newLocation")
    			//			if (newLocation.startsWith("urn:ietf:wg:oauth:2.0:oob")) {
    			//				val code:get
    			//				val title:from
    			//				val accessToken = service.getAccessToken(verifier)
    			//				doSomething(accessToken.getAccessToken())
    			//			}
    		}
    		com.sun.javafx.webkit.WebConsoleListener.setDefaultListener { webview, message, lineNumber, sourceId -> println("Console: [$sourceId:$lineNumber] $message") }
    		engine.setOnError({ event -> System.out.println(event.getMessage()) })
    		try{
    		engine.load(loadURL )
    	} catch (e: IOException) {
    		println("caught error:")
    		e.printStackTrace();
    	}
    }

### PR DESCRIPTION
**Pull request by ararunprasad**
Currently [subresource integrity check](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) is not working in Linux and Windows due to missing implementation of `PAL::CryptoDigest` on those platforms. However it works in macOS because we rely on CommonCrypto based CryptoDigest implementation from WebCore. 

The goal of this PR is to implement `PAL::CryptoDigest` using `java.security.MessageDigest`., so that the subresource integrity can be supported across all the platforms.

After this, `PAL::CryptoDigest` will make use of `java.security.MessageDigest` based implementation on all platforms.

JBS link: https://bugs.openjdk.java.net/browse/JDK-8219917

This PR will fix #230.

### CODE REVIEW

**Comment by muralibilla** [comment]
Is it possible to write a test case for this issue?

**Comment by ararunprasad** [comment]
It needs a HttpServer to test this. I will give a quick try using [JDK-11 HttpServer package](https://docs.oracle.com/en/java/javase/11/docs/api/jdk.httpserver/com/sun/net/httpserver/HttpServer.html).

**Comment by ararunprasad** [comment]
Looks like it is possible to write tests without HttpServer, thanks to https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity.

Adding an integrity field(like below) to script tag should work for our case.

        <script src="sri.js" integrity="sha384-F81A1mZpqritgyD4BB8F8ZttlqrCfKm+BxE+OcAEYPQfukN24x4Y41958fwIBsa2" crossorigin="anonymous"></script>

File: modules/javafx.web/src/main/java/com/sun/javafx/webkit/WCMessageDigestImpl.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+package com.sun.javafx.webkit;
+
+import com.sun.webkit.security.WCMessageDigest;
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class WCMessageDigestImpl extends WCMessageDigest {
+    final private MessageDigest digest;
```
Comment (Maxoudela) [inline modules/javafx.web/src/main/java/com/sun/javafx/webkit/WCMessageDigestImpl.java]: Should this be "private final" according to Java language specification? https://sbforge.org/sonar/rules/show/squid:ModifiersOrderCheck?layout=false

File: modules/javafx.web/src/main/java/com/sun/webkit/security/WCMessageDigest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    protected static WCMessageDigest getInstance(String algorithm) {
+        try {
+            WCMessageDigest digest = new WCMessageDigestImpl(algorithm);
+            return WCMessageDigestPerfLogger.isEnabled() ? new WCMessageDigestPerfLogger(digest) : digest;
+        } catch (Exception ex) {
+            return null;
+        }
+    }
+
+    public abstract void addBytes(ByteBuffer input);
```
Comment (Maxoudela) [inline modules/javafx.web/src/main/java/com/sun/webkit/security/WCMessageDigest.java]: Event if internal, could these abstract methods have some javadoc in order to be more understandable in the future?

**Pull request merged**